Loop/View Models/SimpleBolusViewModel.swift

```diff
@@ -0,0 +1,332 @@
+
+    ///
+    func computeSimpleBolusRecommendation(mealCarbs: HKQuantity?, manualGlucose: HKQuantity?) -> HKQuantity?
+
+    ///
+    var preferredGlucoseUnit: HKUnit { get }
+    
+    ///
+    var maximumBolus: Double { get }
+
+    ///
+    var suspendThreshold: HKQuantity { get }
+}
+
+class SimpleBolusViewModel: ObservableObject {
+    
+    var authenticate: AuthenticationChallenge = LocalAuthentication.deviceOwnerCheck
+
+    enum Alert: Int {
+        case maxBolusExceeded
+        case carbEntryPersistenceFailure
+        case carbEntrySizeTooLarge
+        case manualGlucoseEntryOutOfAcceptableRange
+        case manualGlucoseEntryPersistenceFailure
+        case infoPopup
+    }
+    
+    @Published var activeAlert: Alert?
+    
+    enum Notice: Int {
+        case glucoseBelowSuspendThreshold
+    }
+
+    @Published var activeNotice: Notice?
+
+    var isNoticeVisible: Bool { return activeNotice != nil }    
+
+    @Published var recommendedBolus: String = "0"
+    
+    @Published var enteredCarbAmount: String = "" {
+        didSet {
+            if let enteredCarbs = Self.carbAmountFormatter.number(from: enteredCarbAmount)?.doubleValue, enteredCarbs > 0 {
+                carbs = HKQuantity(unit: .gram(), doubleValue: enteredCarbs)
+            } else {
+                carbs = nil
+            }
+            updateRecommendation()
+        }
+    }
+
+    @Published var enteredGlucoseAmount: String = "" {
+        didSet {
+            if let enteredGlucose = glucoseAmountFormatter.number(from: enteredGlucoseAmount)?.doubleValue {
+                glucose = HKQuantity(unit: delegate.preferredGlucoseUnit, doubleValue: enteredGlucose)
+                if let glucose = glucose, glucose < suspendThreshold {
+                    activeNotice = .glucoseBelowSuspendThreshold
+                } else {
+                    activeNotice = nil
+                }
+            } else {
+                glucose = nil
+            }
+            updateRecommendation()
+        }
+    }
+
+    @Published var enteredBolusAmount: String {
+        didSet {
+            if let enteredBolusAmount = Self.doseAmountFormatter.number(from: enteredBolusAmount)?.doubleValue, enteredBolusAmount > 0 {
+                bolus = HKQuantity(unit: .internationalUnit(), doubleValue: enteredBolusAmount)
+            } else {
+                bolus = nil
+            }
+        }
+    }
+    
+    private var carbs: HKQuantity? = nil
+    private var glucose: HKQuantity? = nil
+    private var bolus: HKQuantity? = nil
+    
+    var glucoseUnit: HKUnit { return delegate.preferredGlucoseUnit }
+    
+    var suspendThreshold: HKQuantity { return delegate.suspendThreshold }
+
+    private var recommendation: Double? = nil {
+        didSet {
+            if let recommendation = recommendation, let recommendationString = Self.doseAmountFormatter.string(from: recommendation) {
```

Comment:
[nit] Setting `bolus` seems a bit round-about here. The `recommendation` is set here (a `Double`), which creates a `recommendationString` (a `String`), which sets `enteredBolusAmount` (a `String`) which (in `didSet`) parses the `String` back to a `Double` to set `bolus`.

Reply:
It's not completely intentional, but is useful. We need both halves of this (recommendation updates the recommendation, which updates the text entry, and user text entry updates the bolus), and get the complete trip for free.